### PR TITLE
Add declarativeNetRequest (DNR) + MV3 examples

### LINKED ISSUE #496: Manifest v3 examples needed
New MV3 examples shall be created and added to the current examples.

### PR DESCRIPTION
**Pull request by Rob--W**
### Description

Three examples demonstrating the use of the declarativeNetRequest API:
- One minimal example demonstrating the use of static DNR rules to block requests.
- One minimal example demonstrating the use of static DNR rules to redirect requests. Also features several MV3 aspects (action, host_permissions, web_accessible_resources).
- A generic example demonstrating how host permissions can be requested and free forms to input DNR rules.

These examples are designed to be cross-browser compatible. In particular, these extensions do not use `background` because there is currently no single manifest that can support both Firefox and Chrome, due to the lack of event page support in Chrome, and the lack of service worker support in Firefox.

Where relevant the examples also include a comparison with MV2.

### Motivation

Need demos to complement the documentation at https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/declarativeNetRequest

### Related issues and pull requests

Relates to #496. While this PR adds some examples, it does not fully resolve #496 because it avoids the use of `background` to be cross-browser-compatible. More examples are needed before we can close #496.

### CODE REVIEW

**Review by rpl — COMMENTED** [review]
The examples in this PR looks pretty good to me, I haven't yet tried them locally but I collected a few comments that I wanted to submit in the meantime (nothing big, mainly nits, typos and a couple of thoughts).

I'll try it out the 3 extensions briefly tomorrow and after that I expect to be able to just sign-off this PR from my side.

File: dnr-dynamic-with-options/README.md
Comment (Rob--W) [reply dnr-dynamic-with-options/README.md:94]: The explanation is meant to be generic, as an extra form of documentation to explain the necessary migration steps. I've added an extra sentence to clarify that this specific example already has both.

File: dnr-dynamic-with-options/options.js
Comment (Rob--W) [reply dnr-dynamic-with-options/options.js:8]: MV2 are not officially supported in Chrome any more, and new submissions are not allowed by the Chrome Web Store. The MV2 section primarily exist to show the difference between MV2 and MV3, so IMO it's not necessary to also explain how to make it compatible with MV2 in Chrome.

File: dnr-dynamic-with-options/options.js
Comment (Rob--W) [reply dnr-dynamic-with-options/options.js:17]: The result is a boolean. And even if it was not, the purpose of the logic here is to ensure that the specified permissions have been granted. Which is done without user prompt if they have already been granted before.

File: dnr-block-only/testpage.html
Comment (rebloor) [inline dnr-block-only/testpage.html:41]:
```suggestion
  because top-level navigation ("main_frame") requests are not matched
```

**Review by rebloor — COMMENTED** [review]
Some suggestions for your consideration

**Review by rpl — APPROVED** [review]
As anticipated in my review pass yesterday, I did try it out each of these 3 extensions locally, followed the instructions on the README to use the example and then adapt it to an MV2 extension that can run on Firefox and provide the exact same behaviors on the MV3 one.

I confirm that the extensions were working as described in the README and that following the steps to adapt to a Firefox MV2 extension also worked as expected on all 3 of them.

As a side note I also tried all 3 MV3 extensions examples on Chrome as well, the only differences in behavior that I've noticed were all highlighted already (e.g. on removing the granted permission in dnr-redirect-url was one, and an inline comment nearby where the errors is reported is already explaining the different behavior in chrome), and all the DNR related behavior worked as expected also on Chrome.

Feel free to merge this once you have integrated all you need from the feedback pass from Richard.

File: dnr-redirect-url/README.md
Comment (Rob--W) [reply dnr-redirect-url/README.md:13]: I created this example first and forgot to do the wrapping. I'll fix this up before merging.

File: dnr-redirect-url/README.md
Comment (Rob--W) [reply dnr-redirect-url/README.md:14]: Yes, the pattern matches http and https.

**Review by Rob--W — COMMENTED** [review]
Thanks for the language improvements Richard. I'll apply almost all of them (and add comments where the suggestion cannot be applied as-is).

Github choked at my attempt to use the UI to apply the changes, so I'll update the PR and force-push the commit.

File: dnr-redirect-url/README.md
Comment (Rob--W) [reply dnr-redirect-url/README.md:14]: I addressed this by changing the sentence to:
```suggestion
- example.com URLs matching regular expression `^https?://([^?]+)$` to the same URL but with the
  scheme set to `https:` (if it was `http:` before), and with `?redirected_by_regex` appended.
```

File: dnr-block-only/README.md
Comment (Rob--W) [reply dnr-block-only/README.md:25]: I changed this sentence to:
```suggestion
This example is the only cross-browser way to block network requests (at least
in Firefox, Chrome, and Safari). The webRequest API is an alternative way to
implement this functionality, but is only available in Firefox (MV2 and MV3)
and in Chrome (MV2 only). Safari does not support the webRequest API.              
```

File: dnr-dynamic-with-options/README.md
Comment (Rob--W) [reply dnr-dynamic-with-options/README.md:69]: Permanently is not a term, that is the normal case. When loaded at `about:debugging` or with `web-ext run`, an extension is loaded temporarily. I rephrased so that the default case is obvious, with the temporary/unloaded state as the exception:

```suggestion
  across extension restarts. These rules also persist across browser restarts,
  unless the extension is loaded temporarily or unloaded.
```

**Pull request merged**
